### ISSUE TITLE
Getting exceptions while calling AcquireTokenByAuthorizationCodeAsync

### ISSUE THREAD

**Issue by huajunzhao** [AzureAD/microsoft-authentication-library-for-dotnet#985]
### Description Of The Issue

Having the below issue while using OpenID authentication while calling AcquireTokenByAuthorizationCodeAsync second time.

AADSTS54005: OAuth2 Authorization code was already redeemed, please retry with a new valid code or use an existing refresh token.
Trace ID: b45bb11c-f9fc-4c0a-aedb-da05f9530f00
Correlation ID: 1ebfc004-ee0c-4fb6-9799-e845ba3322e2
Timestamp: 2019-03-18 14:33:42Z

I researched the above issue, the recommended approach  in the github (https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Acquiring-tokens-with-authorization-codes-on-web-apps ) is to apply context.HandleCodeRedemption, but which seems only available for ASP.NET core, our project is based on ASP.NET, can you please advise how to resolve this issue or if there is any configuration problem of our AAD application?

Sample code

public partial class Startup
    {
        public void ConfigureAuth(IAppBuilder app)
        {
            var clientId = AzureSupportConfiguration.Instance.ClientId;
            var password = SecretConfiguration.Instance.SupportAadPassword;
            var authority = AzureSupportConfiguration.Instance.Authority;
            var redirectUri = AzureSupportConfiguration.Instance.RedirectUri;

         app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);

            app.UseCookieAuthentication(new CookieAuthenticationOptions());

            app.UseOpenIdConnectAuthentication(
                new OpenIdConnectAuthenticationOptions
                {
                    ClientId = clientId,
                    Authority = authority,
                    RedirectUri = redirectUri,

                    TokenValidationParameters = new TokenValidationParameters
                    {
                        // ValidateIssuer set to false to allow personal and work accounts from any organization to sign in to your application
                        // To only allow users from a single organizations, set ValidateIssuer to true and 'tenant' setting in web.config to the tenant name
                        // To allow users from only a list of specific organizations, set ValidateIssuer to true and use ValidIssuers parameter 
                        ValidateIssuer = false,
                    },
                    Notifications = new OpenIdConnectAuthenticationNotifications()
                    {
                        AuthenticationFailed = context =>
                        {
                            context.HandleResponse();
                            //Logger.LogError($"ConfigureAuth.Notifications.AuthenticationFailed:Exception.Message={context.Exception.ToString()}");

                            //var oidException = context.Exception as OpenIdConnectProtocolInvalidNonceException;
                            //if (oidException != null)
                            //{
                            //    // If the NONCE cookie is not set this may just be the user clicking back in the browser and resubmitting
                            //    // the login authorization code.
                            //    // In this case the user likely already has a valid session cookie so we will skip to the next authenticator.
                            //    // If the user dos not have a valid session cookie then they will be signed out and will have to sign in again.
                            //    Logger.LogError($"ConfigureAuth.Notifications.AuthenticationFailed:Exception.Message:Inside.OidException");
                            //    context.SkipToNextMiddleware();
                            //}

                            //var url = string.Format(AzureSupportConfiguration.Instance.ErrorPage, System.Web.HttpUtility.UrlEncode(context.Exception.Message));

                            context.Response.Redirect(AzureSupportConfiguration.Instance.ErrorPage);
                            //context.Response.Redirect(AzureSupportConfiguration.Instance.RedirectUri);

                            return Task.FromResult(0);
                        },
                        RedirectToIdentityProvider = context =>
                        {
                            //context.ProtocolMessage.Prompt = "consent";

                            return Task.FromResult(0);
                        },
                        SecurityTokenReceived = t => Task.FromResult(0),
                        AuthorizationCodeReceived = async context =>
                        {
                            var code = context.Code;
                            var credential = new ClientCredential(clientId, password);
                            var orgId = context.AuthenticationTicket.Identity.Name;
                            var signInUserId = context.AuthenticationTicket.Identity.FindFirst(ClaimTypes.NameIdentifier).Value;
                            var authContext = new AuthenticationContext(authority);

                            var resourceToken = new ResourceToken();
                            var managementTokenResult = await authContext.AcquireTokenByAuthorizationCodeAsync(code, new Uri(redirectUri), credential, AzureSupportConfiguration.Instance.ManagementResource);
                            resourceToken.ManagementToken = managementTokenResult.AccessToken;

                            var expiredOn = managementTokenResult.ExpiresOn;

                            if (AzureSupportConfiguration.Instance.IsGraphApiEnable)
                            {
                               // Logger.LogInfo($"ConfigureAuth.Notifications.AuthorizationCodeReceived:Message.IsGraphApiEnable=true");
                                var graphTokenResult = await authContext.AcquireTokenByAuthorizationCodeAsync(code, new Uri(redirectUri), credential, AzureSupportConfiguration.Instance.GraphResource);
                                resourceToken.GraphToken = graphTokenResult.AccessToken;
                                expiredOn = managementTokenResult.ExpiresOn.CompareTo(graphTokenResult.ExpiresOn) < 0 ? managementTokenResult.ExpiresOn : graphTokenResult.ExpiresOn;
                            }

                            var expiredTimespan = expiredOn.ToUniversalTime() - DateTimeOffset.Now.ToUniversalTime();

                            AadResourceService.SetAccessToken(orgId, resourceToken, expiredTimespan);
                        },
                    }
                });
        }
    }

**Comment by huajunzhao** [comment]
@jmprieur can your team please investigate and help?

**Comment by huajunzhao** [comment]
Just to clarify, i don't find the method context.HandleCodeRedemption(), it seems only exist in ASP.NET core 2.2. https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.openidconnect.authorizationcodereceivedcontext.handlecoderedemption?view=aspnetcore-2.2

Our project is not using ASP.NET core yet, 

![image](https://user-images.githubusercontent.com/3183325/54573768-92969480-4a28-11e9-921e-24bee9dd553a.png)

![image](https://user-images.githubusercontent.com/3183325/54573815-c83b7d80-4a28-11e9-8f6a-3747f02e2785.png)

**Comment by huajunzhao** [comment]
The weird thing is the above code works with an AAD application we registered earlier.
The Application Id: 799fccda-9cb8-41de-8985-28ec1a03c7d5
However it doesn't work with a new AAD application, and throw the AADSTS54005 error.
Application Id:  26f75bc2-aa94-4ce4-901d-5aafa5fa8309

I found it says we can use AcquireTokenSilentAsync after googling this errors.

if (AzureSupportConfiguration.Instance.IsGraphApiEnable)
                            {
                                **var authenticationResult = await authContext.AcquireTokenSilentAsync(AzureSupportConfiguration.Instance.GraphResource, credential, UserIdentifier.AnyUser);**
                                resourceToken.GraphToken = authenticationResult.AccessToken;
                                expiredOn = managementTokenResult.ExpiresOn.CompareTo(authenticationResult.ExpiresOn) < 0 ? managementTokenResult.ExpiresOn : authenticationResult.ExpiresOn;
                            }

it indeed works,.

But we want you guys to help us address the cause why the same code doesn't work with the new AAD app and if this is the right solution?

Thanks,
Kevin

**Comment by jmprieur** [comment]
@huajunzhao 
Azure AD has gone through the OIDC compliance process, and to be OIDC compliant, the auth code should not be redeemable more than once. I would think that this was not applied to previously registered applications (in order to avoid breaking them), but this is definitively enforced for new applications. that's for the explanation

To solve your problem, I believe that in ASP.NET the context passed to the AuthorieationCodeReceived event ha method `HandleResponse()`:
you should just call once you have redeemed the code so that ASP.NET does not do it after you

`context.HandleResponse();`

**Comment by huajunzhao** [comment]
@jmprieur Thanks for the explanation. it makes sense.

however, i attempt to solve this issue by following your suggestion, it still throw the same exception, below are my changes.

![image](https://user-images.githubusercontent.com/3183325/54668460-e76e0400-4b29-11e9-9dbe-c82d78173400.png)

please check and let me know if anything is wrong.

**Comment by jmprieur** [comment]
I think that the problem is that the call to AcquireTokenbyAuthorizationCodeAsync is async (and therefore things probably happen in parallel)
Could you please try to :
- move context.HandleResponse() before the call to AcquireTokenbyAuthorizationCodeAsync. Indeed, even if it failed (it should not), you would not want ASP.NET to redeem the code anyway. 
- or call AcquireTokenbyAuthorizationCodeAsync synchronously (I would not recommend)

**Comment by huajunzhao** [comment]
@jmprieur 
Unfortunately, it doesn't help by moving the context.HandleResponse() before the call to AcquireTokenbyAuthorizationCodeAsync and get the same errors

![image](https://user-images.githubusercontent.com/3183325/54728739-a07c1f00-4bba-11e9-81a1-8098ae822a96.png)

**Comment by huajunzhao** [comment]
Are there any other alternatives? 


As i mentioned earlier, i used the below method to replace the 2nd AcquireTokenbyAuthorizationCodeAsync, it seems work.

 var authenticationResult = await authContext.**AcquireTokenSilentAsync**(AzureSupportConfiguration.Instance.GraphResource, credential, UserIdentifier.AnyUser);

Could you please check if this is a safe workaround?

**Comment by huajunzhao** [comment]
@jmprieur  Any time to look at this?

**Comment by jmprieur** [comment]
Yes, this is fine. you redeem the code once, and then you use AcquireTokenSilent. 
Sorry that I missed the first picture (was using my phone, probably).

Why did you attempt to redeem the code twice ?

**Comment by huajunzhao** [comment]
We're taking over a legacy code, which seems redeem twice against the different resources as below for different usage:

- https://management.chinacloudapi.cn/"
- https://microsoftgraph.chinacloudapi.cn" 

Thanks for your confirmation of using AcquireTokenSilent

however, we start to see the sporadic error while using AcquireTokenSilent
	
outerType
Microsoft.IdentityModel.Clients.ActiveDirectory.AdalSilentTokenAcquisitionException
outerMessage
Failed to acquire token silently as no token was found in the cache. Call method AcquireToken

innermostMessage
{"error":"invalid_grant","error_description":"AADSTS65001: The user or administrator has not consented to use the application with ID '26f75bc2-aa94-4ce4-901d-5aafa5fa8309' named 'acnsupport'. Send an interactive authorization request for this user and resource.\r\nTrace ID: d6341d50-0f9a-461c-ac04-235510e80600\r\nCorrelation ID: f1f34533-c95a-4fea-87c7-5d5828633815\r\nTimestamp: 2019-03-27 05:26:35Z","error_codes":[65001],"timestamp":"2019-03-27 05:26:35Z","trace_id":"d6341d50-0f9a-461c-ac04-235510e80600","correlation_id":"f1f34533-c95a-4fea-87c7-5d5828633815","suberror":"consent_required"}: Unknown error

Do you have any idea?

**Comment by jmprieur** [comment]
@huajunzhao 
Thanks for the update. Indeed, you should not re-redeem the code when you want an access token for a different resource (the OIDC standard says the code should be redeemed only once, and this is being enforced by Azure AD now)

For the consent_required error, is this happening in your Web App? or in another Web API?
If this is happening in a Web App, you might be wanting to have a look at this sample (and this magic line): https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/c34fac62333c3eaac90a9699bdeb16543dd36220/3-WebApp-multi-APIs/Controllers/HomeController.cs#L38

the implementation of what this does is here:
https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/c34fac62333c3eaac90a9699bdeb16543dd36220/Microsoft.Identity.Web/Client/MsalUiRequiredExceptionFilterAttribute.cs#L30

**Comment by huajunzhao** [comment]
@jmprieur  Cool, thanks for the info.

Just want to confirm:

The above code was in ConfigureAuth method in Startup.Auth.cs from ASP.NET web app named " Acn.Portal.Support", below are the main code snippet.

namespace Acn.Portal.Support
{
    public partial class Startup
    {
        //private ILogger logger;
        private TelemetryClient _telemetry;

        public void ConfigureAuth(IAppBuilder app)
        {
            var clientId = AzureSupportConfiguration.Instance.ClientId;
            var password = SecretConfiguration.Instance.SupportAadPassword;
            var authority = AzureSupportConfiguration.Instance.Authority;
            var redirectUri = AzureSupportConfiguration.Instance.RedirectUri;

            app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);

            app.UseCookieAuthentication(new CookieAuthenticationOptions());

            app.UseOpenIdConnectAuthentication(
                new OpenIdConnectAuthenticationOptions
                {
                  ...........
                  Notifications = new OpenIdConnectAuthenticationNotifications()
                    {
                        AuthenticationFailed = context =>
                        {
                            context.HandleResponse();    
                            context.Response.Redirect(AzureSupportConfiguration.Instance.ErrorPage);
                            return Task.FromResult(0);
                        },
                        RedirectToIdentityProvider = context =>
                        {
                            return Task.FromResult(0);
                        },
                        SecurityTokenReceived = t => Task.FromResult(0),
                        AuthorizationCodeReceived = async context =>
                        {
                            var code = context.Code;
                            var credential = new ClientCredential(clientId, password);
                            var orgId = context.AuthenticationTicket.Identity.Name;
                            var signInUserId = context.AuthenticationTicket.Identity.FindFirst(ClaimTypes.NameIdentifier).Value;
                            var authContext = new AuthenticationContext(authority);

                            var resourceToken = new ResourceToken();
                            var managementTokenResult = await authContext.AcquireTokenByAuthorizationCodeAsync(code, new Uri(redirectUri), credential, AzureSupportConfiguration.Instance.ManagementResource);
                            resourceToken.ManagementToken = managementTokenResult.AccessToken;

                            var expiredOn = managementTokenResult.ExpiresOn;

                            if (AzureSupportConfiguration.Instance.IsGraphApiEnable)
                            {
                                var authenticationResult = await authContext.AcquireTokenSilentAsync(AzureSupportConfiguration.Instance.GraphResource, credential, UserIdentifier.AnyUser);
                                resourceToken.GraphToken = authenticationResult.AccessToken;
                                expiredOn = managementTokenResult.ExpiresOn.CompareTo(authenticationResult.ExpiresOn) < 0 ? managementTokenResult.ExpiresOn : authenticationResult.ExpiresOn;
                            }

                            var expiredTimespan = expiredOn.ToUniversalTime() - DateTimeOffset.Now.ToUniversalTime();

                            AadResourceService.SetAccessToken(orgId, resourceToken, expiredTimespan);
                        },
                    }
                });
}

do you mean i add the MsalUiRequiredExceptionFilter attribute for the ConfigureAuth method as well?

**Comment by jmprieur** [comment]
@huajunzhao : no. It should only be added on controller actions.
Did you get the error in the ConfigureAuth method?
if you did, you might be missing a scope.

**Comment by huajunzhao** [comment]
@jmprieur  the “consent_required“” error was happening in ConfigureAuth  method? What do you mean by  "you might be missing a scope.", can you please clarify?

Thanks,
Kevin

**Comment by jmprieur** [comment]
Sure I meant the openid connect scopes
https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/master/Microsoft.Identity.Web/StartupHelpers.cs#L115

**Comment by huajunzhao** [comment]
@jmprieur  thanks for the information.

So what openid connection scopes should I add and how?

Notes: our code is not asp.net core. it was quite old written in asp.net. i have added as below:

![image](https://user-images.githubusercontent.com/3183325/55315740-889c7900-549f-11e9-95d3-e01298fd16ed.png)

Can you pleaes check if this is correct?

Thanks again,
Kevin

**Comment by jmprieur** [comment]
@huajunzhao from what I can see (I mean with the context you provide, and guessing what ResourceToken might be) this looks good.

**Comment by huajunzhao** [comment]
@jmprieur  Thanks for the clarification. I'm testing this change and monitoring. will reach you if i have any further questions.

Thanks for your great support!

**Comment by henrik-me** [comment]
Closing. Please re-open or create an new issue if we can assist.
